src/main/java/carpet/script/utils/ShapeDispatcher.java

```diff
@@ -546,6 +563,129 @@ public long calcKey()
+        float turn;
+        float size;
+        int light_fromblock;
+        int light_fromsky;
+
+        float height;
+        float width;
+        boolean needf3b;
+        CompoundTag blockEntity;
+        BlockState blockState;
+        ItemStack item = null;
+        TransformType item_transform_type;
+
+        @Override
+        protected void init(Map<String, Value> options) {
+            super.init(options);
+            pos = vecFromValue(options.get("pos"));
+            blockState = ((BlockValue) options.getOrDefault("block", BlockValue.AIR)).getBlockState();
+            blockEntity = ((BlockValue) options.getOrDefault("block", BlockValue.AIR)).getData();
+            NBTSerializableValue item_ = (NBTSerializableValue) options.getOrDefault("item", null);
+            if (item_ != null) {
+                this.item = ItemStack.of(item_.getCompoundTag());
+            }
+            light_fromblock = NumericValue
+                    .asNumber(options.getOrDefault("light_fromblock", optional.get("light_fromblock"))).getInt();
+            if (light_fromblock > 15)
+                light_fromblock = 15;
+            light_fromsky = NumericValue.asNumber(options.getOrDefault("light_fromsky", optional.get("light_fromsky")))
+                    .getInt();
+            if (light_fromsky > 15)
+                light_fromsky = 15;
+
+            item_transform_type = TransformType.valueOf(options.getOrDefault("item_transform_type",new StringValue("GUI")).getString());
+
+            String dir = options.getOrDefault("facing", optional.get("facing")).getString();
+            facing = null;
+            switch (dir)
+            {
+                case "north": facing = Direction.NORTH; break;
+                case "south": facing = Direction.SOUTH; break;
+                case "east": facing = Direction.EAST; break;
+                case "west": facing = Direction.WEST; break;
+                case "up":  facing = Direction.UP; break;
+                case "down":  facing = Direction.DOWN; break;
+            }
+
+            needf3b = false;
+            if (options.containsKey("needf3b")) {
+                needf3b = options.get("needf3b").getBoolean();
+            }
+
+            tilt = NumericValue.asNumber(options.getOrDefault("tilt", optional.get("tilt"))).getFloat();
+            lean = NumericValue.asNumber(options.getOrDefault("lean", optional.get("lean"))).getFloat();
+            turn = NumericValue.asNumber(options.getOrDefault("turn", optional.get("turn"))).getFloat();
+            height = NumericValue.asNumber(options.getOrDefault("height", optional.get("height"))).getFloat();
+            width = NumericValue.asNumber(options.getOrDefault("width", optional.get("width"))).getFloat();
+            size = NumericValue.asNumber(options.getOrDefault("obj_size", optional.get("obj_size"))).getFloat();
+        }
+
+        @Override
+        public Consumer<ServerPlayer> alternative() {
+            return s -> {};
```

Comment:
for blocks, technically you can fetch its map color and use that for a particle in the box its rendered on, or display just the block particle, which is vanilla.

Reply:
same for items that are block items

Reply:
> for blocks, technically you can fetch its map color and use that for a particle in the box its rendered on, or display just the block particle, which is vanilla.

i get its map color to implent its alternative.